### PR TITLE
Support referencing stack slots in the DWARF debug info

### LINKED ISSUE #3884: Debugging module with lldb shows local variables as not available
### Test Case

[test.wasm.zip](https://github.com/bytecodealliance/wasmtime/files/8189246/test.wasm.zip)

```c
#include <stdio.h>

int main() {
    int i;
    for(i = 0; i < 5; i++) {
        printf("%d\n", i);
    }
    return 0;
}
```
compiled with: `wasi-sdk-14.0/bin/clang -g -O0 test.c -o test.wasm`

### Steps to Reproduce

```
> lldb -- wasmtime run -g test.wasm                  
(lldb) target create "wasmtime"
Current executable set to 'wasmtime' (arm64).
(lldb) settings set -- target.run-args  "run" "-g" "test.wasm"
(lldb) settings set plugin.jit-loader.gdb.enable on
(lldb) b test.c:6
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run
Process 40146 launched: '/Users/jeffcharles/.cargo/bin/wasmtime' (arm64)
1 location added to breakpoint 1
Process 40146 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x0000000100e940e0 JIT(0x1102b0000)`main at test.c:6:24
   3   	int main() {
   4   	    int i;
   5   	    for(i = 0; i < 5; i++) {
-> 6   	        printf("%d\n", i);
   7   	    }
   8   	    return 0;
   9   	}
Target 0: (wasmtime) stopped.
(lldb) fr v
(WasmtimeVMContext *) __vmctx = <variable not available>

(int) i = <no location, value may have been optimized out>
```

### Expected Results

Running `fr v` in lldb in this context should display:
```
(int) i = 0
```

### Actual Results

```
(int) i = <no location, value may have been optimized out>
```

### Versions and Environment

Wasmtime version or commit: 0.34.0

Operating system: MacOS Monterrey (12.2.1)

Architecture: arm64

### Extra Info

LLDB version:
```
> lldb --version
lldb-1300.0.42.3
Swift version 5.5.2-dev
```

This is what happens when I try to compile and run natively to show what I'm expecting:
```
> clang -g -O0 test.c -o test
> lldb -- test
(lldb) target create "test"
Current executable set to '/Users/jeffcharles/projects/wasm32-wasi/test-c-app/test' (arm64).
(lldb) b test.c:6
Breakpoint 1: where = test`main + 32 at test.c:6:24, address = 0x0000000100003f6c
(lldb) run
Process 40384 launched: '/Users/jeffcharles/projects/wasm32-wasi/test-c-app/test' (arm64)
Process 40384 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x0000000100003f6c test`main at test.c:6:24
   3   	int main() {
   4   	    int i;
   5   	    for(i = 0; i < 5; i++) {
-> 6   	        printf("%d\n", i);
   7   	    }
   8   	    return 0;
   9   	}
Target 0: (test) stopped.
(lldb) fr v
(int) i = 0
```

### PR DESCRIPTION
**Pull request by SingleAccretion**
Implements the plan essentially outlined in https://github.com/bytecodealliance/wasmtime/issues/2856#issuecomment-1192419659:

1) Emit or use SysV-style unwind info if debug info is requested.
2) Reference that unwind info in the form of `.debug_frame` using `DW_OP_call_frame_cfa`.
3) Plumb the right offsets through.

Fixes #3884 (at least to some degree - however, that issue does not contain anything actionable), and also the example I was looking at in C#, though only for `-opt-level 0` codegen.

### CODE REVIEW

File: crates/cranelift/src/debug/write_debuginfo.rs [hunk 1]
```diff
@@ -146,7 +149,14 @@ fn create_frame_table<'a>(
     let cie_id = table.add_cie(isa.create_systemv_cie()?);
 
     for (i, metadata) in funcs {
+        let mut unwind_info: Option<&CfaUnwindInfo> = None;
         if let Some(UnwindInfo::SystemV(info)) = &metadata.unwind_info {
+            unwind_info = Some(info);
+        } else if let Some(info) = &metadata.cfa_unwind_info {
+            unwind_info = Some(info);
+        }
```
Comment (alexcrichton) [inline crates/cranelift/src/debug/write_debuginfo.rs:157]: This makes me pause for a bit for a few reasons. One is could this assert that at most one of these fields are `SystemV` unwind information? Hopefully to head off any possible future confusion.

The second reason is that in Wasmtime [only `unwind_info` is consulted](https://github.com/bytecodealliance/wasmtime/blob/9377dfd7b82c8cc5bfdfbc435df1b633b2506f08/crates/cranelift-shared/src/obj.rs#L132-L134) meaning that it seems to be like we wouldn't emit the DWARF unwinding information? It also seems though like you're probably running/testing on Windows, so I'm curious if you are how this is working?

File: tests/all/debug/testsuite/spilled_frame_base.wasm
Comment (alexcrichton) [inline tests/all/debug/testsuite/spilled_frame_base.wasm:1]: I realize that this is preexisting, so there's no need to fix this, but ideally we wouldn't  check in wasm binaries for tests. In the ideal world CI would have a pinned version of wasi-sdk that it downloads and uses to compile these guest programs (as part of the test). 

In lieu of this though could you perhaps comment in the test itself how to generate the wasm file? For example what C flags were passed. That's probably sufficient for now and a future commit could come through to delete the wasm files and build them on CI instead.

**Review by alexcrichton — COMMENTED** [review]
Looks all reasonable to me, thanks again for working on this! I'm no expert myself in debuginfo but I don't think we necessarily have a resident expert to turn to, so happy to review at a somewhat higher level and lean on you for the DWARF correctness and all that.

One final comment I'd have in addition to those below is that our current debuginfo tests are pretty primitive, so if you're interested or so inspired I think it'd be great to soup them up. For example I'd love to have "just" a directory of C and/or Rust files which are compiled automatically in CI and the source itself has various special comments for commands/breakpoints/etc.

File: cranelift/codegen/src/machinst/vcode.rs [hunk 2]
```diff
@@ -1220,6 +1217,21 @@ impl<I: VCodeInst> VCode<I> {
...
+                    trace!(
+                        "Extending debug range for VL{} in {:?} to {}",
+                        label,
+                        loc,
+                        end
+                    );
+                    last_loc_range.end = end;
+                    continue;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (SingleAccretion) [reply cranelift/codegen/src/machinst/vcode.rs:1233]: Correct.

File: crates/cranelift/src/debug/write_debuginfo.rs (hunk 1 above)
Comment (SingleAccretion) [reply crates/cranelift/src/debug/write_debuginfo.rs:157]: > Could this assert that at most one of these fields are SystemV unwind information?

Sure, will add. It's not really _incorrect_ for them to be `{SysV, SysV}`, just wasteful in terms of memory.

> I'm curious how this is working?

This function is a callee of `append_dwarf`, which itself adds to the final image. So on Windows we will have `.debug_frame`, much like the other DWARF sections.

It is interesting to note what happens when we are on Unix - in that case the unwind info will be effectively emitted twice (note this is pre-existing behavior). AFAICT from the code the "unwind" part will be emitted into `.eh_frame`, so they don't conflict.

In effect, the "debug info" use of this information is completely decoupled from the "unwind" use.

File: tests/all/debug/testsuite/spilled_frame_base.wasm
Comment (SingleAccretion) [reply tests/all/debug/testsuite/spilled_frame_base.wasm:1]: Agree, binaries in source control seem like a large question mark. Will add the arguments.

**Comment by alexcrichton** [comment]
Oh interesting! So there's no way for the dwarf sections to actually refer to `.eh_frame`? It's required at the binary level that the information is duplicated?

**Comment by SingleAccretion** [comment]
> So there's no way for the dwarf sections to actually refer to .eh_frame? It's required at the binary level that the information is duplicated?

I am no great Unix expert myself, so do not know for certain. I _suspect_ that it is indeed possible for the debugger to use `.eh_frame` alone, though the two sections are a little different in format details as well as usage (`.eh_frame` is like `.pdata` on Windows in that it is "mandatory").

FWIW, there are much larger (pun intended) size problems with `-g`-produced object files at the moment. For one, all the location lists for variables with are very fragmented, it would help a lot if they could always refer to the one and only stack location, at least for unoptimized code.

**Review by alexcrichton — APPROVED** [review]
Ok and no worries, I wanted to confirm for my own curiosity mostly. I think it's reasonable to basically incrementally improve things and any work in this space is very much appreciated!

**Comment by bjorn3** [comment]
Rustc doesn't ever emit `.debug_frame`. The debugger has to use `.eh_frame` for this information for rust code.

**Pull request merged**